### ISSUE TITLE
Update traffic join to avoid 'traffic evaporating'

### ISSUE THREAD

**Issue by Robinlovelace** [nptscot/npt#555]
There are small anomalies popping up e.g.:

![image](https://github.com/user-attachments/assets/04039214-3a04-41f9-b7a2-c9ae73d8af1d)

[https://osm.org/go/evfLeWTk?m=](https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fosm.org%2Fgo%2FevfLeWTk%3Fm%3D&data=05%7C02%7Cr.lovelace%40leeds.ac.uk%7C8fab5e02479d469b525d08dd0fbd18b4%7Cbdeaeda8c81d45ce863e5232a535b7cb%7C0%7C0%7C638684030307139053%7CUnknown%7CTWFpbGZsb3d8eyJFbXB0eU1hcGkiOnRydWUsIlYiOiIwLjAuMDAwMCIsIlAiOiJXaW4zMiIsIkFOIjoiTWFpbCIsIldUIjoyfQ%3D%3D%7C0%7C%7C%7C&sdata=qpTbWlaPbpzcFJL2cVy71OIX%2FieGM6d2cYuZ6AIKQ28%3D&reserved=0)

It appears if the traffic is evaporating between these two high-flow section. There were a couple of other examples of this that occur if you need me to send them across too.

**Comment by Robinlovelace** [comment]
Source: output of this:

```
> newhaven = zonebuilder::zb_zone("newhaven edinburgh")
> traffic_newhaven = traffic_volumes_scotland[newhaven, ]
nrow(traffic_> nrow(traffic_newhaven)
[1] 22672
> newhaven = zonebuilder::zb_zone("newhaven edinburgh", n_circles = 1)
Please set segment_center = TRUE to divide the centre into multiple segments
> traffic_newhaven = traffic_volumes_scotland[newhaven, ]
> mapview::mapview(traffic_newhaven, zcol = "pred_flows")
Browsing http://127.0.0.1:7996
```

**Comment by Robinlovelace** [comment]
But the segment is totally missing from the CBD layer:

![image](https://github.com/user-attachments/assets/066f7642-f146-4569-9399-5b6cc728263f)

Source: https://nptscot.github.io/#/clos/#14.21/55.97816/-3.2004

**Comment by wangzhao0217** [comment]
error stem from this line:
```
osm_district = osm_national |>
        filter(osm_id %in% district_centroids$osm_id)
```
![image](https://github.com/user-attachments/assets/8936c057-a9f3-4cb2-896d-4bdeffa841f8)

**Comment by wangzhao0217** [comment]
Can be fix by using:
```
      osm_district = osm_national[sf::st_union(district_geom), , op = sf::st_intersects]
```
but it is slow

![image](https://github.com/user-attachments/assets/f1fa377f-ccff-47e9-88f9-ef7e4afadab9)

**Comment by Robinlovelace** [comment]
Great diagnosis of the issue, thanks Zhao!

**Comment by Robinlovelace** [comment]
> Can be fix by using:
> 
> ```
>       osm_district = osm_national[sf::st_union(district_geom), , op = sf::st_intersects]
> ```

Another issue with that solution: it will lead to duplicate IDs. A reason for using centroids is you only get 1 per region.

**Comment by Robinlovelace** [comment]
I'm on the case: I think I can come up with a quick fix to the LADs.

**Comment by Robinlovelace** [comment]
Fix in progress:

![image](https://github.com/user-attachments/assets/79618922-8155-4eb4-9f64-36c6f188fb7a)

**Comment by Robinlovelace** [comment]
That section is mostly fixed:

![image](https://github.com/user-attachments/assets/71cbb34a-0673-4150-966e-b6bc09f5a601)

**Comment by Robinlovelace** [comment]
Source: https://nptscot.github.io/#/clos/#13.9/55.98072/-3.20461

**Comment by Robinlovelace** [comment]
Still a tiny bit missing: 

![image](https://github.com/user-attachments/assets/ed7d3266-38bf-4d43-83f6-799e118b4e80)

**Comment by Robinlovelace** [comment]
Source: https://nptscot.github.io/#/clos/#15.84/55.980376/-3.202827

**Comment by Robinlovelace** [comment]
Should be fixed upstream in https://github.com/nptscot/osmactive/pull/66

**Comment by Robinlovelace** [comment]
Fixed in #550
